File: src/VBNetFx/FAC_Generic/LicenseManager/FacessoLicenseManager.vb
Imports ActiveDev
Imports Microsoft.Win32

''' <summary>
''' Übernimmt das komplette Licencing der Facesso-Software. Von AdLicenseManager abgeleitet.
''' </summary>
''' <remarks></remarks>
<CLSCompliant(False)> _
Public NotInheritable Class FacessoLicenseManager
    Inherits ADLicenseManager

    Sub New(ByVal prgGuid As Guid, ByVal InstallDate As Date, ByVal LastRunDate As Date, ByVal LastRegisteredDate As Date, ByVal SerialNumber As String)
        MyBase.New(prgGuid, InstallDate, LastRunDate, LastRegisteredDate, SerialNumber)
        If Not HasValidSerialNo() Then
            myLicenseInfo.Fallback(1, 2)
        End If
    End Sub

    Public Overrides Function IsLicensed() As Boolean
        If Date.Now < myInstallDate Then
            Try
                'Wir müssen auf CURRENT_USER auf jeden Fall setzen, und...
                Registry.SetValue("HKEY_CURRENT_USER\SOFTWARE\ActiveDev\Facesso", _
                        "ForceReapplication", True)

                '...LOCAL_MACHINE in jedem Fall versuchen (klappt unter 64Bit nur als Admin).
                Registry.SetValue("HKEY_LOCAL_MACHINE\SOFTWARE\ActiveDev\Facesso", _
                        "ForceReapplication", True)

            Catch ex As Exception
            End Try
            Dim up As New ADLicenseUnvalidException(Date.Now, ADLicenseUnvalidReason.SystemDateManipulated, _
                "Das aktuelle Datum liegt vor dem Installationsdatum. Facesso kann daher die korrekten Lizenzinformationen nicht überprüfen und wird deswegen nun beendet. Beim nächsten Start von Facesso können Sie einen neuen Freischaltcode eingeben!")
            Throw up
            Return False
        End If

        If Date.Now < myLastRunDate Then
            Try
                'Wir müssen auf CURRENT_USER auf jeden Fall setzen, und...
                Registry.SetValue("HKEY_CURRENT_USER\SOFTWARE\ActiveDev\Facesso", _
                        "ForceReapplication", True)

                '...LOCAL_MACHINE in jedem Fall versuchen (klappt unter 64Bit nur als Admin).
                Registry.SetValue("HKEY_LOCAL_MACHINE\SOFTWARE\ActiveDev\Facesso", _
                        "ForceReapplication", True)

            Catch ex As Exception
            End Try
            Dim up As New ADLicenseUnvalidException(Date.Now, ADLicenseUnvalidReason.SystemDateManipulated, _
                "Das aktuelle Datum liegt vor dem letzten Startdatum von Facesso. Eine Überprüfung der korrekten Lizenzinformationen ist daher nicht möglich. Beim nächsten Start von Facesso können Sie einen neuen Freischaltcode eingeben.")
            Throw up
            Return False
        End If

        If myLicenseInfo.MonthsLimited <> 0 Then
            If Date.Now > BestBefore Then
                Try
                    'Wir müssen auf CURRENT_USER auf jeden Fall setzen, und...
                    Registry.SetValue("HKEY_CURRENT_USER\SOFTWARE\ActiveDev\Facesso", _
                            "ForceReapplication", True)

                    '...LOCAL_MACHINE in jedem Fall versuchen (klappt unter 64Bit nur als Admin).
                    Registry.SetValue("HKEY_LOCAL_MACHINE\SOFTWARE\ActiveDev\Facesso", _
                            "ForceReapplication", True)

                Catch ex As Exception
                End Try
                Dim up As New ADLicenseUnvalidException(Date.Now, ADLicenseUnvalidReason.Expired, _
                    "Die Probe-Frist ist abgelaufen! Beim nächsten Start von Facesso können Sie einen neuen Freischaltcode eingeben!")
                Throw up
                Return False
            End If
        End If

        If myLicenseInfo.SoftwareID > 10 Then
            Dim up As New ADLicenseUnvalidException(Date.Now, ADLicenseUnvalidReason.WrongSoftware, _
                "Der Freischaltcode ist nicht für diese Softwareversion gültig! Beim nächsten Start von Facesso können Sie einen neuen Freischaltcode eingeben.")
            Throw up
            Return False
        End If

        If myLicenseInfo.HasFallenBack Then
            Return True
        Else
            Return MyBase.IsLicensed
        End If
    End Function

    Public ReadOnly Property VersionPermissionInfo() As FacessoVersionPermissionInfo
        Get
            Return New FacessoVersionPermissionInfo(CType(myLicenseInfo.SoftwareID, FacessoVersion))
        End Get
    End Property

    Public Overrides Function ToString() As String
        Dim tmpIsLicenced As Boolean
        Dim retText As String = "Ist Lizensiert: " & If(IsLicensed, "Ja", "Nein") & vbNewLine
        tmpIsLicenced = tmpIsLicenced Or IsLicensed()
        retText &= "Has Fallen Back: " & If(myLicenseInfo.HasFallenBack, "Ja", "Nein") & vbNewLine
        retText &= "Beschränkt auf Monate: " & myLicenseInfo.MonthsLimited
        tmpIsLicenced = tmpIsLicenced Or myLicenseInfo.HasFallenBack

        If Not tmpIsLicenced Then Return retText

        retText &= vbNewLine & "Software-ID: " & myLicenseInfo.SoftwareID.ToString & vbNewLine
        retText &= "Für Nutzer: " & myLicenseInfo.Limit1.ToString & vbNewLine
        retText &= "Für Internet-Nutzer: " & myLicenseInfo.Limit2.ToString & vbNewLine
        retText &= "Für Mitarbeiter: " & myLicenseInfo.Limit3.ToString & vbNewLine
        retText &= "Schnittstellenausbau : " & myLicenseInfo.Limit4.ToString & vbNewLine
        Return retText
    End Function
End Class

File: src/MixNetFx/FAC_Generic/LicenseManager/FacessoLicenseManager.cs
using System;
using ActiveDev;
using Microsoft.Win32;

namespace Facesso
{
    [CLSCompliant(false)]
    public sealed class FacessoLicenseManager : ADLicenseManager
    {
        public FacessoLicenseManager(Guid prgGuid, DateTime installDate, DateTime lastRunDate,
                                     DateTime lastRegisteredDate, string serialNumber)
            : base(prgGuid, installDate, lastRunDate, lastRegisteredDate, serialNumber)
        {
            if (!HasValidSerialNo())
                myLicenseInfo.Fallback(1, 2);
        }

        public override bool IsLicensed()
        {
            if (DateTime.Now < myInstallDate)
            {
                try
                {
                    Registry.SetValue(@"HKEY_CURRENT_USER\SOFTWARE\ActiveDev\Facesso", "ForceReapplication", true);
                    Registry.SetValue(@"HKEY_LOCAL_MACHINE\SOFTWARE\ActiveDev\Facesso", "ForceReapplication", true);
                }
                catch { }
                throw new ADLicenseUnvalidException(DateTime.Now, ADLicenseUnvalidReason.SystemDateManipulated,
                    "Das aktuelle Datum liegt vor dem Installationsdatum. Facesso kann daher die korrekten Lizenzinformationen nicht überprüfen und wird deswegen nun beendet. Beim nächsten Start von Facesso können Sie einen neuen Freischaltcode eingeben!");
            }

            if (DateTime.Now < myLastRunDate)
            {
                try
                {
                    Registry.SetValue(@"HKEY_CURRENT_USER\SOFTWARE\ActiveDev\Facesso", "ForceReapplication", true);
                    Registry.SetValue(@"HKEY_LOCAL_MACHINE\SOFTWARE\ActiveDev\Facesso", "ForceReapplication", true);
                }
                catch { }
                throw new ADLicenseUnvalidException(DateTime.Now, ADLicenseUnvalidReason.SystemDateManipulated,
                    "Das aktuelle Datum liegt vor dem letzten Startdatum von Facesso. Eine Überprüfung der korrekten Lizenzinformationen ist daher nicht möglich. Beim nächsten Start von Facesso können Sie einen neuen Freischaltcode eingeben.");
            }

            if (myLicenseInfo.MonthsLimited != 0)
            {
                if (DateTime.Now > BestBefore)
                {
                    try
                    {
                        Registry.SetValue(@"HKEY_CURRENT_USER\SOFTWARE\ActiveDev\Facesso", "ForceReapplication", true);
                        Registry.SetValue(@"HKEY_LOCAL_MACHINE\SOFTWARE\ActiveDev\Facesso", "ForceReapplication", true);
                    }
                    catch { }
                    throw new ADLicenseUnvalidException(DateTime.Now, ADLicenseUnvalidReason.Expired,
                        "Die Probe-Frist ist abgelaufen! Beim nächsten Start von Facesso können Sie einen neuen Freischaltcode eingeben!");
                }
            }

            if (myLicenseInfo.SoftwareID > 10)
            {
                throw new ADLicenseUnvalidException(DateTime.Now, ADLicenseUnvalidReason.WrongSoftware,
                    "Der Freischaltcode ist nicht für diese Softwareversion gültig! Beim nächsten Start von Facesso können Sie einen neuen Freischaltcode eingeben.");
            }

            if (myLicenseInfo.HasFallenBack)
                return true;
            else
                return base.IsLicensed();
        }

        public FacessoVersionPermissionInfo VersionPermissionInfo
        {
            get { return new FacessoVersionPermissionInfo((FacessoVersion)myLicenseInfo.SoftwareID); }
        }

        public override string ToString()
        {
            bool tmpIsLicenced = false;
            string retText = "Ist Lizensiert: " + (IsLicensed() ? "Ja" : "Nein") + "\r\n";
            tmpIsLicenced = tmpIsLicenced | IsLicensed();
            retText += "Has Fallen Back: " + (myLicenseInfo.HasFallenBack ? "Ja" : "Nein") + "\r\n";
            retText += "Beschränkt auf Monate: " + myLicenseInfo.MonthsLimited;
            tmpIsLicenced = tmpIsLicenced | myLicenseInfo.HasFallenBack;

            if (!tmpIsLicenced) return retText;

            retText += "\r\n" + "Software-ID: " + myLicenseInfo.SoftwareID.ToString() + "\r\n";
            retText += "Für Nutzer: " + myLicenseInfo.Limit1.ToString() + "\r\n";
            retText += "Für Internet-Nutzer: " + myLicenseInfo.Limit2.ToString() + "\r\n";
            retText += "Für Mitarbeiter: " + myLicenseInfo.Limit3.ToString() + "\r\n";
            retText += "Schnittstellenausbau : " + myLicenseInfo.Limit4.ToString() + "\r\n";
            return retText;
        }
    }
}
